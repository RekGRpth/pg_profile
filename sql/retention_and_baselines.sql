UPDATE profile.samples
SET sample_time = now() - (5 - sample_id) * '1 day'::interval - '10 minutes'::interval
WHERE sample_id <= 5;

ALTER TABLE profile.test_rel_storage_params1 RESET (autovacuum_vacuum_threshold,fillfactor,autovacuum_enabled);
ALTER INDEX profile.ix_test_rel_storage_params_id1 RESET (fillfactor);
ALTER INDEX profile.ix_test_rel_storage_params_val1 RESET (fillfactor);

SELECT server,result FROM profile.take_sample();

SELECT relname, t.last_sample_id, reloptions
FROM profile.table_storage_parameters t JOIN profile.tables_list
  USING (server_id, relid)
WHERE relname IN ('test_rel_storage_params1','test_rel_storage_params2') ORDER BY relid, last_sample_id;
SELECT relname, indexrelname, i.last_sample_id, reloptions
FROM profile.index_storage_parameters i JOIN profile.tables_list
  USING (server_id, relid)
  JOIN profile.indexes_list USING (server_id, relid, indexrelid)
WHERE relname IN ('test_rel_storage_params1','test_rel_storage_params2') ORDER BY relid, indexrelid, last_sample_id;

BEGIN;
  SELECT profile.delete_samples();
  SELECT sample FROM profile.show_samples() ORDER BY sample;
ROLLBACK;
SELECT count(*) FROM profile.samples WHERE sample_time < now() - '1 days'::interval;
SELECT * FROM profile.set_server_max_sample_age('local',1);
/* Testing baseline creation */
SELECT * FROM profile.create_baseline('testline1',2,4);
BEGIN;
  SELECT profile.delete_samples('local',tstzrange(
      (SELECT sample_time FROM profile.samples WHERE sample_id = 1),
      (SELECT sample_time FROM profile.samples WHERE sample_id = 5),
      '[]'
    )
  );
  SELECT sample FROM profile.show_samples() ORDER BY sample;
ROLLBACK;
BEGIN;
  SELECT profile.delete_samples(tstzrange(
      (SELECT sample_time FROM profile.samples WHERE sample_id = 1),
      (SELECT sample_time FROM profile.samples WHERE sample_id = 5),
      '[]'
    )
  );
  SELECT sample FROM profile.show_samples() ORDER BY sample;
ROLLBACK;
SELECT * FROM profile.create_baseline('testline2',2,4);
SELECT count(*) FROM profile.baselines;
SELECT * FROM profile.keep_baseline('testline2',-1);
/* Testing baseline show */
SELECT baseline, min_sample, max_sample, keep_until_time IS NULL
FROM profile.show_baselines()
ORDER BY baseline;
/* Testing baseline deletion */
SELECT server,result FROM profile.take_sample();
SELECT count(*) FROM profile.baselines;
/* Testing samples retention override with baseline */
SELECT count(*) FROM profile.samples WHERE sample_time < now() - '1 days'::interval;
SELECT * FROM profile.drop_baseline('testline1');
/* Testing samples deletion after baseline removed */
SELECT server,result FROM profile.take_sample();
SELECT count(*) FROM profile.samples WHERE sample_time < now() - '1 days'::interval;
