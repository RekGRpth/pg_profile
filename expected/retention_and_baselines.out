UPDATE profile.samples
SET sample_time = now() - (5 - sample_id) * '1 day'::interval - '10 minutes'::interval
WHERE sample_id <= 5;
ALTER TABLE profile.test_rel_storage_params1 RESET (autovacuum_vacuum_threshold,fillfactor,autovacuum_enabled);
ALTER INDEX profile.ix_test_rel_storage_params_id1 RESET (fillfactor);
ALTER INDEX profile.ix_test_rel_storage_params_val1 RESET (fillfactor);
SELECT server,result FROM profile.take_sample();
 server | result 
--------+--------
 local  | OK
(1 row)

SELECT relname, t.last_sample_id, reloptions
FROM profile.table_storage_parameters t JOIN profile.tables_list
  USING (server_id, relid)
WHERE relname IN ('test_rel_storage_params1','test_rel_storage_params2') ORDER BY relid, last_sample_id;
         relname          | last_sample_id |                                         reloptions                                          
--------------------------+----------------+---------------------------------------------------------------------------------------------
 test_rel_storage_params1 |              8 | ["autovacuum_vacuum_threshold=50", "fillfactor=100", "autovacuum_enabled=true"]
 test_rel_storage_params2 |                | ["vacuum_index_cleanup=true", "vacuum_truncate=true", "autovacuum_vacuum_scale_factor=0.2"]
(2 rows)

SELECT relname, indexrelname, i.last_sample_id, reloptions
FROM profile.index_storage_parameters i JOIN profile.tables_list
  USING (server_id, relid)
  JOIN profile.indexes_list USING (server_id, relid, indexrelid)
WHERE relname IN ('test_rel_storage_params1','test_rel_storage_params2') ORDER BY relid, indexrelid, last_sample_id;
         relname          |          indexrelname           | last_sample_id |     reloptions     
--------------------------+---------------------------------+----------------+--------------------
 test_rel_storage_params1 | ix_test_rel_storage_params_id1  |              8 | ["fillfactor=100"]
 test_rel_storage_params1 | ix_test_rel_storage_params_val1 |              8 | ["fillfactor=90"]
 test_rel_storage_params2 | ix_test_rel_storage_params_id2  |                | ["fillfactor=80"]
 test_rel_storage_params2 | ix_test_rel_storage_params_val2 |                | ["fillfactor=70"]
(4 rows)

BEGIN;
  SELECT profile.delete_samples();
 delete_samples 
----------------
              8
(1 row)

  SELECT sample FROM profile.show_samples() ORDER BY sample;
 sample 
--------
      9
(1 row)

ROLLBACK;
SELECT count(*) FROM profile.samples WHERE sample_time < now() - '1 days'::interval;
 count 
-------
     4
(1 row)

SELECT * FROM profile.set_server_max_sample_age('local',1);
 set_server_max_sample_age 
---------------------------
                         1
(1 row)

/* Testing baseline creation */
SELECT * FROM profile.create_baseline('testline1',2,4);
 create_baseline 
-----------------
               1
(1 row)

BEGIN;
  SELECT profile.delete_samples('local',tstzrange(
      (SELECT sample_time FROM profile.samples WHERE sample_id = 1),
      (SELECT sample_time FROM profile.samples WHERE sample_id = 5),
      '[]'
    )
  );
 delete_samples 
----------------
              2
(1 row)

  SELECT sample FROM profile.show_samples() ORDER BY sample;
 sample 
--------
      2
      3
      4
      6
      7
      8
      9
(7 rows)

ROLLBACK;
BEGIN;
  SELECT profile.delete_samples(tstzrange(
      (SELECT sample_time FROM profile.samples WHERE sample_id = 1),
      (SELECT sample_time FROM profile.samples WHERE sample_id = 5),
      '[]'
    )
  );
 delete_samples 
----------------
              2
(1 row)

  SELECT sample FROM profile.show_samples() ORDER BY sample;
 sample 
--------
      2
      3
      4
      6
      7
      8
      9
(7 rows)

ROLLBACK;
SELECT * FROM profile.create_baseline('testline2',2,4);
 create_baseline 
-----------------
               2
(1 row)

SELECT count(*) FROM profile.baselines;
 count 
-------
     2
(1 row)

SELECT * FROM profile.keep_baseline('testline2',-1);
 keep_baseline 
---------------
             1
(1 row)

/* Testing baseline show */
SELECT baseline, min_sample, max_sample, keep_until_time IS NULL
FROM profile.show_baselines()
ORDER BY baseline;
 baseline  | min_sample | max_sample | ?column? 
-----------+------------+------------+----------
 testline1 |          2 |          4 | t
 testline2 |          2 |          4 | f
(2 rows)

/* Testing baseline deletion */
SELECT server,result FROM profile.take_sample();
 server | result 
--------+--------
 local  | OK
(1 row)

SELECT count(*) FROM profile.baselines;
 count 
-------
     1
(1 row)

/* Testing samples retention override with baseline */
SELECT count(*) FROM profile.samples WHERE sample_time < now() - '1 days'::interval;
 count 
-------
     3
(1 row)

SELECT * FROM profile.drop_baseline('testline1');
 drop_baseline 
---------------
             1
(1 row)

/* Testing samples deletion after baseline removed */
SELECT server,result FROM profile.take_sample();
 server | result 
--------+--------
 local  | OK
(1 row)

SELECT count(*) FROM profile.samples WHERE sample_time < now() - '1 days'::interval;
 count 
-------
     0
(1 row)

